packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,199 @@
+    if (this.props.shapes !== prevProps.shapes) {
+      this.clearAllShapes()
+      this.addShape(
+        this.formatLongLatToGmapsCoordinates(
+          this.props.shapes
+        )[0]
+      )
+    }
+  }
+
+  getPolygonCoords(poly) {
+    const len = poly.getPath().getLength()
+    const shape = []
+
+    for (let i = 0; i < len; i += 1) {
+      shape.push([poly.getPath().getAt(i).lng(), poly.getPath().getAt(i).lat()])
+    }
+    return shape
+  }
+
+  polygons = []
+
+  enableMapControls() {
+    const { map, mapControls } = this.props
+    map.setOptions({
+      draggable: true,
+      zoomControl: true,
+      scrollwheel: true,
+      disableDoubleClickZoom: true,
+      ...mapControls,
+    })
+    this.setState({
+      drawing: false,
+    })
+  }
+
+  disableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: false,
+      zoomControl: false,
+      scrollwheel: false,
+      disableDoubleClickZoom: false,
+    })
+    this.setState({
+      drawing: true,
+    })
+  }
+
+  @autobind
+  enableMapDraw() {
+    const { map, multipleShapes } = this.props
+    this.disableMapControls()
+
+    if (!multipleShapes) {
+      this.clearAllShapes()
+    }
+    setupEvents(map, EVENTS, {
+      onMouseDown: () => {
+        this.drawFreeHand()
+      },
+    }, true)
+  }
+
+  drawFreeHand() {
+    const { map, onMapDrawStart, dataStyle, onMapDrawEnd } = this.props
+    const poly = new window.google.maps.Polyline({
+      map,
+      clickable: false,
+      ...dataStyle.polyline,
+    })
+    const drawPoly = new window.google.maps.Polygon({
+      map,
+      path: poly.getPath(),
+      clickable: false,
+      ...dataStyle.polylineFill,
+    })
+
+    if (onMapDrawStart) {
+      onMapDrawStart()
+    }
+
+    const listener = setupEvents(map, EVENTS, {
+      onMouseMove: e => {
+        poly.getPath().push(e.latLng)
+      },
+      onTouchStart: e => {
+        poly.getPath().push(e.latLng)
+      },
+    }, false)
+
+    const onMouseUp = () => {
+      removeEvent(listener.onMouseMove)
+      poly.setMap(null)
+      drawPoly.setMap(null)
+      if (onMapDrawEnd) {
+        onMapDrawEnd(this.getPolygonCoords(poly))
+      }
+      this.enableMapControls()
+    }
+
+    setupEvents(map, EVENTS, {
+      onMouseUp,
+      onTouchEnd: onMouseUp,
+    }, true)
+  }
+
+  addShapeToGMaps(path) {
+    const { map, dataStyle } = this.props
+    return new window.google.maps.Polygon({
+      map,
+      path,
+      ...dataStyle.polygon,
+    })
+  }
+
+  addShape(path) {
+    const rawPolygon = this.addShapeToGMaps(path)
+    this.polygons = [...this.polygons, rawPolygon]
+  }
+
+  @autobind
+  clearAllShapes() {
+    this.polygons.forEach(poly => poly.setMap(null))
+    this.polygons = []
+  }
+
+  formatLongLatToGmapsCoordinates(shapes) {
+    // Should be refactored later to
+    // over(pipe(allArr, allArr), point => ({lat: point[1], lng: point[0]})(Object.values(shape)
```

Comment:
Why can't we do this now?

Reply:
We need to add Ramda

Reply:
@orionstein suggested it on Friday and he wanted us to talk to you about installing ramda and using it. I think he wanted to have a library with all the lenses/over/pipe ramda functions we'd need before installing it though.

Reply:
Yeah, completely 💯 on ramda.